File: sdk/android/core/src/main/kotlin/com/iapstack/core/IapStackErrors.kt
package com.iapstack.core

sealed class IapStackException(message: String, cause: Throwable? = null) : Exception(message, cause)

/**
 * Structured v1 API error returned by IAPStack.
 */
class IapStackApiException(
  val statusCode: Int,
  val code: String,
  val requestId: String? = null,
  val retryable: Boolean,
  message: String,
) : IapStackException(message)

/**
 * Network-level transport error before a full API response exists.
 */
class IapStackTransportException(
  message: String,
  cause: Throwable? = null,
) : IapStackException(message, cause)

/**
 * One bounded HTTP attempt exceeded its deadline or was aborted.
 */
class IapStackTimeoutException(
  message: String,
  cause: Throwable? = null,
) : IapStackException(message, cause)

/**
 * Response did not match the contract shape.
 */
class IapStackProtocolException(
  message: String,
  cause: Throwable? = null,
) : IapStackException(message, cause)

File: sdk/ios/Sources/IAPStackApple/IAPStackErrors.swift
import Foundation

/// Structured IAPStack errors returned or produced by this SDK.
public enum IAPStackSDKError: Error, LocalizedError, Sendable {
  /// Structured API error returned by IAPStack with a machine-readable status.
  case apiError(statusCode: Int, code: String, message: String, requestId: String?, retryable: Bool)

  /// Network-level transport error before an HTTP payload is decoded.
  case transportError(message: String, cause: (any Error)? = nil)

  /// Per-attempt request timeout.
  case timeoutError(message: String, cause: (any Error)? = nil)

  /// Contract, JSON, or response-shape mismatch.
  case protocolError(message: String, cause: (any Error)? = nil)

  /// Invalid user-provided configuration.
  case configurationError(message: String)

  public var errorDescription: String? {
    switch self {
    case let .apiError(statusCode: statusCode, code: code, message: message, requestId: requestId, retryable: _):
      return "IAPStack API error (status: \(statusCode), code: \(code), requestId: \(requestId ?? "nil"), message: \(message))"
    case let .transportError(message: message, cause: cause):
      return "IAPStack transport error: \(message)\(cause.flatMap({ ": \($0.localizedDescription)" }) ?? "")"
    case let .timeoutError(message: message, cause: cause):
      return "IAPStack request timeout: \(message)\(cause.flatMap({ ": \($0.localizedDescription)" }) ?? "")"
    case let .protocolError(message: message, cause: cause):
      return "IAPStack protocol error: \(message)\(cause.flatMap({ ": \($0.localizedDescription)" }) ?? "")"
    case let .configurationError(message):
      return "IAPStack configuration error: \(message)"
    }
  }

  public var isRetryable: Bool {
    switch self {
    case let .apiError(_, _, _, _, retryable):
      return retryable
    case .protocolError, .configurationError:
      return false
    case .transportError, .timeoutError:
      return true
    }
  }
}
